vmware/terraform-provider-vcd
# Add leftovers removal and clean up tests

* Add leftovers removal to test suite (it will run automatically)
    * `vcd/config_test.go`
    * `vcd/remove_leftovers_test.go` (new file)
    * `scripts/get_test_resource_names.go` (new file)
    * `scripts/runtest.sh`
    * `TESTING.md`
* Remove check for existing catalog before and after suite (it was added before we had a way of populating VCDs with automated operations)
    * `vcd/config_test.go` 
* Remove redundant `// +build` tags, which are not needed since we use Go 1.19+ by default
    * All files `vcd/*_.test.go` 

New testing options:

```
-vcd-silent-leftovers-removal              Omit details during removal of leftovers (false)
-vcd-skip-leftovers-removal                Do not attempt removal of leftovers at the end of the test suite (false)
```

**NOTES:**

1. The feature removal handles the following resources, which are the most likely ones to be left behind:
    * `vcd_org`
    * `vcd_org_vdc`
    * `vcd_catalog`
    * `vcd_catalog_vapp_template`
    * `vcd_catalog_media`
    * `vcd_network_*`
    * `vcd_vapp`
    * `vcd_vm`

2. The file `scripts/get_resource_names.go` gets copied to `./vcd/test-artifacts` during `make test-binary`, and can be used to generate a list of the resources created during tests. There shouldn't be a need to run it again, except when we notice that some test resources are left behind after a failed test.

3. to run the removal without another test run (for example, when the failing test was running in another machine) execute the following:

```
go test -tags functional -v -run NoSuchTest # or any other non-existing test name
```
